File: sample-app-hd-calculator/src/main/java/com/hd/calculator/app/function/db/AppDatabase.java
package com.hd.calculator.app.function.db;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import com.hd.calculator.app.function.db.dao.AccountDao;
import com.hd.calculator.app.function.db.dao.BillDailyReportDao;
import com.hd.calculator.app.function.db.dao.BillRecordDao;
import com.hd.calculator.app.function.db.dao.DishesDao;
import com.hd.calculator.app.function.db.dao.DishesSortDao;
import com.hd.calculator.app.function.db.dao.OrderRecordDao;
import com.hd.calculator.app.function.db.dao.ReduceDishesRecordDao;
import com.hd.calculator.app.function.db.dao.TableDao;
import com.hd.calculator.app.function.db.dao.TaxDao;
import com.hd.calculator.app.function.db.entity.AccountEntity;
import com.hd.calculator.app.function.db.entity.DishesEntity;
import com.hd.calculator.app.function.db.entity.DishesSortEntity;
import com.hd.calculator.app.function.db.entity.TableEntity;
import com.hd.calculator.app.function.db.entity.TaxEntity;
import com.hd.calculator.app.function.db.entity.operation.BillDailyReportEntity;
import com.hd.calculator.app.function.db.entity.operation.BillDishesRecordEntity;
import com.hd.calculator.app.function.db.entity.operation.BillRecordEntity;
import com.hd.calculator.app.function.db.entity.operation.OrderDishesRecordEntity;
import com.hd.calculator.app.function.db.entity.operation.OrderRecordEntity;
import com.hd.calculator.app.function.db.entity.operation.ReduceDishesRecordEntity;


@Database(entities = {
        AccountEntity.class,
        DishesEntity.class,
        DishesSortEntity.class,
        TaxEntity.class,
        TableEntity.class,
        OrderRecordEntity.class,
        OrderDishesRecordEntity.class,
        BillRecordEntity.class,
        BillDishesRecordEntity.class,
        ReduceDishesRecordEntity.class,
        BillDailyReportEntity.class},
        version = 1,
        exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {

    // 单例实例
    private static volatile AppDatabase INSTANCE;

    //getDatabase
    public static AppDatabase getDatabase(final Context context) {
        if (INSTANCE == null) {
            synchronized (AppDatabase.class) {
                if (INSTANCE == null) {
                    INSTANCE = Room.databaseBuilder(context.getApplicationContext(), AppDatabase.class, "app_database.db").fallbackToDestructiveMigration() // 开发阶段使用（生产环境移除）
                            .build();
                }
            }
        }
        return INSTANCE;
    }

    public abstract AccountDao accountDao();

    //dishesDao
    public abstract DishesSortDao dishesSortDao();

    public abstract DishesDao dishesDao();

    public abstract TaxDao taxDao();

    //tableDao
    public abstract TableDao tableDao();

    public abstract OrderRecordDao orderRecordDao();

    public abstract BillRecordDao billRecordDao();

    public abstract ReduceDishesRecordDao reduceDishesRecordDao();

    public abstract BillDailyReportDao billDailyReportDao();
}

File: sample-im/src/main/java/app/allever/android/sample/im/database/AppDatabase.kt
package app.allever.android.sample.im.database

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import app.allever.android.sample.im.database.dao.UserDao
import app.allever.android.sample.im.database.entity.UserEntity

@Database(entities = [UserEntity::class], version = 2, exportSchema = false)
abstract class AppDatabase : RoomDatabase() {

    abstract fun userDao(): UserDao

    companion object {
        @Volatile
        private var INSTANCE: AppDatabase? = null

        // 1->2 迁移：新增 online 字段
        private val MIGRATION_1_2 = object : Migration(1, 2) {
            override fun migrate(database: SupportSQLiteDatabase) {
                database.execSQL("ALTER TABLE user ADD COLUMN online INTEGER NOT NULL DEFAULT 0")
            }
        }

        fun init(context: Context) {
            if (INSTANCE == null) {
                synchronized(this) {
                    if (INSTANCE == null) {
                        INSTANCE = Room.databaseBuilder(
                            context.applicationContext,
                            AppDatabase::class.java,
                            "local_im_db"
                        )
                            .addMigrations(MIGRATION_1_2)
                            .build()
                    }
                }
            }
        }

        fun getInstance(): AppDatabase {
            return INSTANCE ?: throw IllegalStateException("请先在 Application 中初始化 AppDatabase")
        }
    }
}